dbms/src/Storages/TrivialStorageBuffer.cpp

```diff
@@ -0,0 +1,550 @@
+    }
+
+    /** If sources from destination table are already processed to non-starting stage, then we should wrap
+      * sources from the buffer to the same stage of processing conveyor.
+      */
+    if (processed_stage > QueryProcessingStage::FetchColumns)
+        for (auto & stream : streams_from_buffers)
+            stream = InterpreterSelectQuery(query, context, processed_stage, 0, stream).execute().in;
+
+    streams.insert(streams.end(), streams_from_buffers.begin(), streams_from_buffers.end());
+    return streams;
+}
+
+void TrivialStorageBuffer::addBlock(const Block & block)
+{
+    SipHash hash;
+    block.updateHash(hash);
+    HashType block_hash = hash.get64();
+
+    std::lock_guard<std::mutex> lock(mutex);
+    if (current_hashes->find(block_hash) == current_hashes->end()
+        && previous_hashes->find(block_hash) == previous_hashes->end())
+    {
+        if (current_hashes->size() >= num_blocks_to_deduplicate / 2)
+        {
+            previous_hashes = std::move(current_hashes);
+            current_hashes = std::make_unique<DeduplicationBuffer>();
+        }
+        current_hashes->insert(block_hash);
+        current_rows += block.rows();
+        current_bytes += block.bytes();
+        data.push_back(block);
+
+        CurrentMetrics::add(CurrentMetrics::StorageBufferRows, current_rows);
+        CurrentMetrics::add(CurrentMetrics::StorageBufferBytes, current_bytes);
+    }
+    else
+    {
+        auto it = previous_hashes->find(block_hash);
+        if (it != previous_hashes->end())
+        {
+            current_hashes->insert(it);
+            previous_hashes->erase(it);
+        }
+    }
+}
+
+void TrivialStorageBuffer::flush(bool check_thresholds)
+{
+    Block block_to_write;
+    time_t current_time = time(0);
+
+    time_t time_passed = 0;
+
+    if (data.empty())
+        return;
+
+    {
+        std::lock_guard<std::mutex> lock(mutex);
+
+        if (first_write_time)
+            time_passed = current_time - first_write_time;
+
+        if (check_thresholds)
+        {
+            if (!checkThresholdsImpl(current_rows, current_bytes, time_passed))
+                return;
+        }
+        else
+        {
+            if (current_rows == 0)
+                return;
+        }
+
+        /// Collecting BlockList into single block.
+        block_to_write = data.front().cloneEmpty();
+        block_to_write.checkNumberOfRows();
+        for (auto & block : data)
+        {
+            block.checkNumberOfRows();
+            for (size_t column_no = 0, columns = block.columns(); column_no < columns; ++column_no)
+            {
+                IColumn & col_to = *block_to_write.safeGetByPosition(column_no).column.get();
+                const IColumn & col_from = *block.getByName(col_to.getName()).column.get();
+
+                col_to.insertRangeFrom(col_from, 0, block.rows());
+            }
+
```

Comment:
You could grab range of blocks from the list under lock.
Then release the lock and merge that range to single block.
New blocks could be added to the end of the list.

After writing to destination, you remove range from the list, instead of clear.

Also you need to be sure, that flush method is not called from multiple threads simultaneously.